### DISCUSSION TITLE
How do I use Rc inside middleware?

### DISCUSSION THREAD

**Discussion by squeaktoy** [tokio-rs/axum#3019] [Q&A] [answered]
### Summary

Hey fellow axumates. I'm working on a project at https://git.sr.ht/~latex/streamlist
The xrust library uses Rc for its RNode object that I'm trying to use in a middleware.
This is not allowed because middlewares are Send.
Do any of you know any ideas how I can work around this issue?
Here are some ideas I and fellow Rustaceans at ##rust on libera.chat came up with:
1. Make xrust thread-safe. I opened a bug report for xrust https://github.com/ballsteve/xrust/issues/123
2. Spawn a new thread that handles all interactions with xrust
3. Somehow make the middleware not Send, e.g. by using Tokio in non-work-stealing mode.

Preferably I don't want to use yet another thread just for xrust. Is it even possible to make the middleware non-Send? Does axum even support that or not? What about actix-web?

### axum version

0.7.7

**Comment by jplatte** [comment, marked as answer]
Using a non-`Send` type inside of an `async fn` does not automatically make its future type non-`Send`. That only happens when you declare a binding of such a type before the last `.await` point (roughly speaking).

If you don't need to hold the `RNode` across a `.await` point, the best solution is probably to move all the `xrust` logic into a regular `fn` that is called by your `async fn`. I took a brief look at your code and think this should work there.

**Reply by squeaktoy** [reply]
So I think I might have fixed the Rc thingy. But using the request inside the middleware is still a mystery for me, but I opened a new discussion for that: https://github.com/tokio-rs/axum/discussions/3022
